File: 93-bounded-context/ContextDemo.java
/*
 * Ngôn ngữ: Java
 * Công dụng: Bounded context — cùng chữ "khách hàng", ba nghĩa khác nhau ở ba đội. Bài
 * cho nổ ba con bug của "một mô hình dùng chung cho cả công ty": đội bán hàng không tạo
 * nổi một khách tiềm năng vì thiếu mã số thuế; chữ "hoàn tất" mang hai nghĩa nên một
 * trong hai đội luôn đọc sai; và mỗi lần thêm một field là ba đội phải cùng lên lịch.
 * Tại sao cần học: đây là bài học kiến trúc đắt nhất trong tầng này, vì nó đi ngược
 * trực giác. Trực giác nói "đừng lặp lại chính mình — một khách hàng thì phải có MỘT
 * lớp KhachHang". Thực tế nói ngược lại: hai đội nói hai ngôn ngữ khác nhau về cùng một
 * con người, và ép họ dùng chung một lớp không loại bỏ sự khác nhau đó — nó chỉ giấu sự
 * khác nhau vào trong những field mà nửa số đội phải bỏ trống.
 */
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class ContextDemo {

    // =====================================================================
    // SAI — MỘT mô hình dùng chung cho cả công ty
    // =====================================================================
    static final class KhachHangChung {
        // Bán hàng cần:
        final String maKhach, ten, nguonKhach, giaiDoanBan;
        // Kế toán cần:
        final String tenPhapNhan, maSoThue, diaChiXuatHoaDon, dieuKhoanThanhToan;
        // Hỗ trợ cần:
        final String email, hangUuTien, ngonNguGiaoTiep;
        // Và một chữ mà cả ba đội cùng dùng, mỗi đội hiểu một kiểu:
        final boolean hoanTat;

        KhachHangChung(String maKhach, String ten, String nguonKhach, String giaiDoanBan,
                       String tenPhapNhan, String maSoThue, String diaChiXuatHoaDon,
                       String dieuKhoanThanhToan, String email, String hangUuTien,
                       String ngonNguGiaoTiep, boolean hoanTat) {
            // Kế toán yêu cầu mã số thuế bắt buộc — hoàn toàn hợp lý VỚI KẾ TOÁN.
            if (maSoThue == null || maSoThue.isBlank())
                throw new IllegalArgumentException("mã số thuế là bắt buộc");
            this.maKhach = maKhach; this.ten = ten; this.nguonKhach = nguonKhach;
            this.giaiDoanBan = giaiDoanBan; this.tenPhapNhan = tenPhapNhan;
            this.maSoThue = maSoThue; this.diaChiXuatHoaDon = diaChiXuatHoaDon;
            this.dieuKhoanThanhToan = dieuKhoanThanhToan; this.email = email;
            this.hangUuTien = hangUuTien; this.ngonNguGiaoTiep = ngonNguGiaoTiep;
            this.hoanTat = hoanTat;
        }
    }

    // =====================================================================
    // ĐÚNG — mỗi bounded context một mô hình, nối nhau BẰNG MÃ
    // =====================================================================

    /** Ngữ cảnh BÁN HÀNG. "Khách hàng" ở đây là một CƠ HỘI đang được theo đuổi. */
    static final class BanHang {
        record KhachHang(String maKhach, String ten, String nguonKhach, GiaiDoan giaiDoan) {
            KhachHang {
                if (ten == null || ten.isBlank())
                    throw new IllegalArgumentException("khách tiềm năng phải có tên");
            }
            /** Với BÁN HÀNG, "hoàn tất" nghĩa là ĐÃ CHỐT ĐƠN. */
            boolean daHoanTat() { return giaiDoan == GiaiDoan.DA_CHOT; }
        }
        enum GiaiDoan { TIEM_NANG, DANG_TU_VAN, DA_CHOT, DA_MAT }
    }

    /** Ngữ cảnh KẾ TOÁN. "Khách hàng" ở đây là một PHÁP NHÂN xuất hoá đơn được. */
    static final class KeToan {
        record BenNhanHoaDon(String maKhach, String tenPhapNhan, String maSoThue,
                             String diaChiXuatHoaDon, boolean daThuDuTien) {
            BenNhanHoaDon {
                if (maSoThue == null || maSoThue.isBlank())
                    throw new IllegalArgumentException("bên nhận hoá đơn phải có mã số thuế");
            }
            /** Với KẾ TOÁN, "hoàn tất" nghĩa là ĐÃ THU ĐỦ TIỀN. */
            boolean daHoanTat() { return daThuDuTien; }
        }
    }

    /** Ngữ cảnh HỖ TRỢ. "Khách hàng" ở đây là một NGƯỜI có thể mở phiếu hỗ trợ. */
    static final class HoTro {
        record NguoiDung(String maKhach, String email, int mucUuTien) { }
    }

    // ---- Self-check ----
    private static void check(boolean ok, String msg) {
        if (!ok) throw new AssertionError(msg);
    }

    public static void main(String[] args) {
        // ---- 1. CON BUG: đội bán hàng không tạo nổi một khách tiềm năng ----
        // 9h sáng ở hội chợ. Nhân viên bán hàng gặp một người, có tên và số điện thoại,
        // muốn ghi lại ngay. Người đó chưa phải khách, chưa có công ty, chưa có mã số thuế.
        boolean biChan = false;
        try {
            new KhachHangChung("KH-01", "Chị Hoa ở hội chợ", "hội chợ", "TIEM_NANG",
                    null, null, null, null, null, null, null, false);
        } catch (IllegalArgumentException e) { biChan = true; }
        check(biChan, "không tạo được: mô hình chung đòi mã số thuế");
        // Luật "mã số thuế là bắt buộc" HOÀN TOÀN ĐÚNG — với kế toán. Nó chỉ sai khi bị
        // áp lên một ngữ cảnh mà khái niệm "khách hàng" còn chưa có nghĩa đó.
        //
        // Cách vá mà mọi dự án đều làm, và vì sao nó tệ hơn: đặt `maSoThue` thành tuỳ
        // chọn. Thế là kế toán mất luôn bảo đảm "mọi bên nhận hoá đơn đều có mã số thuế",
        // và phải tự kiểm ở mọi chỗ dùng. Một ràng buộc thật vừa biến thành lời khuyên.

        BanHang.KhachHang ch = new BanHang.KhachHang(
                "KH-01", "Chị Hoa ở hội chợ", "hội chợ", BanHang.GiaiDoan.TIEM_NANG);
        check(ch.ten().equals("Chị Hoa ở hội chợ"), "ngữ cảnh bán hàng: tạo được với 4 field");
        check(BanHang.KhachHang.class.getRecordComponents().length == 4, "và chỉ có 4 field");
        check(KhachHangChung.class.getDeclaredFields().length == 12, "mô hình chung có 12");
        // Bốn so với mười hai. Tám field kia không phải "dữ liệu chưa điền" — chúng là
        // dữ liệu KHÔNG CÓ NGHĨA trong ngữ cảnh này.

        // ---- 2. CON BUG: cùng một chữ, hai nghĩa ----
        // Đơn của chị Hoa: đã chốt bán (bán hàng gọi là "hoàn tất"), nhưng công nợ 30
        // ngày nên chưa thu tiền (kế toán KHÔNG gọi là hoàn tất).
        BanHang.KhachHang daChot = new BanHang.KhachHang(
                "KH-01", "Chị Hoa", "hội chợ", BanHang.GiaiDoan.DA_CHOT);
        KeToan.BenNhanHoaDon chuaThu = new KeToan.BenNhanHoaDon(
                "KH-01", "Công ty Hoa Mai", "0301234567", "12 Lê Lợi", false);

        check(daChot.daHoanTat(), "BÁN HÀNG: hoàn tất = đã chốt -> ĐÚNG");
        check(!chuaThu.daHoanTat(), "KẾ TOÁN: hoàn tất = đã thu tiền -> CHƯA");
        check(daChot.daHoanTat() != chuaThu.daHoanTat(),
                "cùng một khách, cùng một chữ, hai câu trả lời — và cả hai đều đúng");
        // Với mô hình chung, `hoanTat` là MỘT boolean. Ai gán nó? Đội nào gán thì đội kia
        // đọc sai. Không có cách vá nào ngoài việc tách ra thành hai khái niệm — và tách
        // ra thì đã là hai bounded context rồi.
        //
        // Đây là con bug đúng như bài 81 phần 1, nhưng ở quy mô tổ chức: ở đó là hai lập
        // trình viên hiểu khác nhau, ở đây là hai PHÒNG BAN hiểu khác nhau. Và họ đều đúng.

        // ---- 3. NỐI HAI NGỮ CẢNH BẰNG MÃ, KHÔNG BẰNG OBJECT ----
        check(daChot.maKhach().equals(chuaThu.maKhach()), "cùng một con người ngoài đời");
        // Nhưng KHÔNG có phương thức nào đi từ `BanHang.KhachHang` sang
        // `KeToan.BenNhanHoaDon`. Đây chính là bài 83 (tham chiếu bằng id) nâng lên cấp
        // độ tổ chức: hai ngữ cảnh chia sẻ một ĐỊNH DANH, không chia sẻ một MÔ HÌNH.
        //
        // Và trình biên dịch canh giúp: hai lớp cùng tên `KhachHang` ở hai gói khác nhau
        // là HAI KIỂU khác nhau. Dòng dưới không biên dịch được:
        //     KeToan.BenNhanHoaDon x = daChot;
        //     error: incompatible types: BanHang.KhachHang cannot be converted to ...
        check(!BanHang.KhachHang.class.equals(KeToan.BenNhanHoaDon.class),
                "cùng khái niệm ngoài đời, hai kiểu dữ liệu — và đó là điều ta MUỐN");

        // ---- 4. ĐO CHI PHÍ THAY ĐỔI ----
        // Kế toán cần thêm `dieuKhoanThanhToan`.
        //   Mô hình chung: sửa lớp -> ba đội cùng biên dịch lại, cùng test lại, cùng
        //                  triển khai. Muốn ra bản vá thì phải xếp lịch với hai đội không
        //                  liên quan gì tới thay đổi này.
        //   Tách ngữ cảnh: sửa `KeToan.BenNhanHoaDon` -> đúng một đội, đúng một lần triển khai.
        int doiBiAnhHuongMoHinhChung = 3;
        int doiBiAnhHuongKhiTach = 1;
        check(doiBiAnhHuongMoHinhChung == 3 && doiBiAnhHuongKhiTach == 1, "3 đội so với 1");
        // Đây là con số thật sự quyết định. Mô hình chung không làm code chậm đi — nó làm
        // TỔ CHỨC chậm đi, và đó là thứ đắt hơn nhiều.

        // ---- 5. BẢN ĐỒ NGỮ CẢNH: quan hệ giữa các ngữ cảnh có TÊN ----
        //
        //   Quan hệ            | Nghĩa                                  | Khi nào dùng
        //   -------------------|----------------------------------------|-------------------
        //   Đối tác            | hai đội cùng đổi, cùng chịu trách nhiệm| hai đội cùng công ty
        //   Khách/Nhà cung cấp | thượng nguồn nghe hạ nguồn             | có quyền thương lượng
        //   Tuân thủ           | hạ nguồn dùng y nguyên mô hình trên    | bên trên không đổi được
        //   Chống hư hỏng      | hạ nguồn DỊCH mô hình trên sang của mình| mô hình trên xấu (bài 94)
        //   Nhân chung         | hai đội cùng sở hữu một phần mã dùng chung | rất ít, rất nguy hiểm
        //
        // "Nhân chung" là thứ mọi người bắt đầu và hối hận: một thư viện `common-model`
        // mà ba đội cùng sửa. Nó có mọi nhược điểm của mô hình chung, cộng thêm việc
        // không ai sở hữu nó.
        Map<String, String> banDo = new LinkedHashMap<>();
        banDo.put("BanHang -> KeToan", "Khách/Nhà cung cấp: bán hàng chốt đơn, kế toán xuất hoá đơn");
        banDo.put("KeToan -> CongThue", "Tuân thủ: cơ quan thuế không đổi định dạng vì ta");
        banDo.put("HoTro -> BanHang", "Chống hư hỏng: hỗ trợ tự dịch, không phụ thuộc giai đoạn bán");
        check(banDo.size() == 3, "bản đồ ngữ cảnh là một tài liệu THẬT, vẽ được trên một trang giấy");
        // Nếu không vẽ được bản đồ này cho hệ thống của bạn, thì ranh giới ngữ cảnh chưa
        // tồn tại — chỉ có các gói code cùng dùng chung một mô hình.

        // ---- 6. DỊCH Ở BIÊN, MỖI CHIỀU MỘT LẦN ----
        // Khi bán hàng chốt đơn, kế toán cần một bên nhận hoá đơn. Việc DỊCH nằm ở biên,
        // và nó nhận thêm dữ liệu mà ngữ cảnh nguồn không có (mã số thuế do kế toán thu thập).
        List<KeToan.BenNhanHoaDon> soHoaDon = new ArrayList<>();
        if (daChot.daHoanTat()) {
            soHoaDon.add(new KeToan.BenNhanHoaDon(daChot.maKhach(), "Công ty Hoa Mai",
                    "0301234567", "12 Lê Lợi", false));
        }
        check(soHoaDon.size() == 1, "dịch ở biên: một chiều, một chỗ, có tên");
        check(soHoaDon.get(0).maKhach().equals(daChot.maKhach()), "chỉ MÃ đi qua biên");
        // Chỗ dịch này là nơi DUY NHẤT hai ngôn ngữ gặp nhau, nên nó là nơi duy nhất phải
        // sửa khi một bên đổi. Bài 94 nói kỹ về việc bảo vệ mình khi bên kia có mô hình xấu.

        // ---- 7. KHI NÀO KHÔNG TÁCH NGỮ CẢNH ----
        // Bounded context có chi phí thật: mô hình lặp lại, mã dịch ở biên, dữ liệu đồng
        // bộ trễ. Ba dấu hiệu cho thấy CHƯA nên tách:
        //   - Cả hệ thống do MỘT đội làm, và mọi người dùng cùng một bộ từ ngữ.
        //   - Chưa tìm ra được một từ nào mang hai nghĩa (phép thử ở phần 2).
        //   - Số field mà mỗi bên phải bỏ trống còn nhỏ.
        // Ngược lại, ba dấu hiệu ĐÃ đến lúc tách:
        //   - Có field mà nửa số nơi dùng luôn để `null`;
        //   - Có từ mà bạn phải hỏi lại "ý anh là hoàn tất theo nghĩa nào";
        //   - Một thay đổi nhỏ phải xếp lịch với đội không liên quan.
        check(KhachHangChung.class.getDeclaredFields().length
                > BanHang.KhachHang.class.getRecordComponents().length * 2,
                "mô hình chung phình gấp ba — dấu hiệu rõ nhất và đo được nhất");

        System.out.println("OK");
    }
}

File: 93-bounded-context/ContextDemo.cpp
/*
 * Ngôn ngữ: C++
 * Công dụng: Bản C++ của cùng bài học — một mô hình dùng chung cho cả công ty chặn đội
 * bán hàng tạo khách tiềm năng, làm chữ "hoàn tất" mang hai nghĩa, và biến mọi thay đổi
 * nhỏ thành việc của ba đội.
 * Tại sao cần học: ở Java và Python, cái giá của "mô hình dùng chung" là chi phí TỔ CHỨC
 * — ba đội phải xếp lịch với nhau. Ở C++ có thêm một cái giá đo được bằng đồng hồ: mô
 * hình chung sống trong một HEADER, và mọi đơn vị biên dịch `#include` nó sẽ được dịch
 * lại mỗi khi nó đổi một dòng. Ranh giới ngữ cảnh ở C++ vì thế không chỉ là chuyện thiết
 * kế, nó là chuyện thời gian build — và đó là thứ cả đội cảm thấy mỗi ngày.
 */
#include <iostream>
#include <map>
#include <stdexcept>
#include <string>
#include <vector>
#include <cstdlib>

// =====================================================================
// SAI — MOT mo hinh dung chung cho ca cong ty
// =====================================================================
struct KhachHangChung {
    // Ban hang can:
    std::string maKhach, ten, nguonKhach, giaiDoanBan;
    // Ke toan can:
    std::string tenPhapNhan, maSoThue, diaChiXuatHoaDon, dieuKhoanThanhToan;
    // Ho tro can:
    std::string email, hangUuTien, ngonNguGiaoTiep;
    // Va mot chu ma ca ba doi cung dung, moi doi hieu mot kieu:
    bool hoanTat = false;

    static KhachHangChung tao(std::string maKhach, std::string ten, std::string maSoThue) {
        // Ke toan yeu cau ma so thue bat buoc — hoan toan hop ly VOI KE TOAN.
        if (maSoThue.empty()) throw std::invalid_argument("ma so thue la bat buoc");
        KhachHangChung k;
        k.maKhach = std::move(maKhach);
        k.ten = std::move(ten);
        k.maSoThue = std::move(maSoThue);
        return k;
    }
};

// =====================================================================
// DUNG — moi bounded context mot mo hinh, noi nhau BANG MA
// =====================================================================

// Ngu canh BAN HANG. "Khach hang" o day la mot CO HOI dang duoc theo duoi.
namespace banHang {
enum class GiaiDoan { TIEM_NANG, DANG_TU_VAN, DA_CHOT, DA_MAT };

struct KhachHang {
    std::string maKhach, ten, nguonKhach;
    GiaiDoan giaiDoan;

    KhachHang(std::string ma, std::string t, std::string nguon, GiaiDoan gd)
        : maKhach(std::move(ma)), ten(std::move(t)), nguonKhach(std::move(nguon)), giaiDoan(gd) {
        if (ten.empty()) throw std::invalid_argument("khach tiem nang phai co ten");
    }
    // Voi BAN HANG, "hoan tat" nghia la DA CHOT DON.
    bool daHoanTat() const { return giaiDoan == GiaiDoan::DA_CHOT; }
};
}  // namespace banHang

// Ngu canh KE TOAN. "Khach hang" o day la mot PHAP NHAN xuat hoa don duoc.
namespace keToan {
struct BenNhanHoaDon {
    std::string maKhach, tenPhapNhan, maSoThue, diaChiXuatHoaDon;
    bool daThuDuTien;

    BenNhanHoaDon(std::string ma, std::string tpn, std::string mst, std::string dc, bool thu)
        : maKhach(std::move(ma)), tenPhapNhan(std::move(tpn)), maSoThue(std::move(mst)),
          diaChiXuatHoaDon(std::move(dc)), daThuDuTien(thu) {
        if (maSoThue.empty()) throw std::invalid_argument("ben nhan hoa don phai co ma so thue");
    }
    // Voi KE TOAN, "hoan tat" nghia la DA THU DU TIEN.
    bool daHoanTat() const { return daThuDuTien; }
};
}  // namespace keToan

// Ngu canh HO TRO. "Khach hang" o day la mot NGUOI co the mo phieu ho tro.
namespace hoTro {
struct NguoiDung {
    std::string maKhach, email;
    int mucUuTien;
};
}  // namespace hoTro

// ---- Self-check ----
static void check(bool ok, const std::string& msg) {
    if (!ok) {
        std::cerr << "FAIL: " << msg << "\n";
        std::exit(1);
    }
}

int main() {
    // ---- 1. CON BUG: doi ban hang khong tao noi mot khach tiem nang ----
    // 9h sang o hoi cho. Nhan vien ban hang gap mot nguoi, co ten va so dien thoai, muon
    // ghi lai ngay. Nguoi do chua phai khach, chua co cong ty, chua co ma so thue.
    bool biChan = false;
    try { KhachHangChung::tao("KH-01", "Chi Hoa o hoi cho", ""); }
    catch (const std::invalid_argument&) { biChan = true; }
    check(biChan, "khong tao duoc: mo hinh chung doi ma so thue");
    // Luat "ma so thue la bat buoc" HOAN TOAN DUNG — voi ke toan. No chi sai khi bi ap
    // len mot ngu canh ma khai niem "khach hang" con chua co nghia do.
    //
    // Cach va ma moi du an deu lam, va vi sao no te hon: dat `maSoThue` thanh tuy chon.
    // The la ke toan mat luon bao dam "moi ben nhan hoa don deu co ma so thue", va phai
    // tu kiem o moi cho dung. Mot rang buoc that vua bien thanh loi khuyen.

    banHang::KhachHang ch("KH-01", "Chi Hoa o hoi cho", "hoi cho", banHang::GiaiDoan::TIEM_NANG);
    check(ch.ten == "Chi Hoa o hoi cho", "ngu canh ban hang: tao duoc voi 4 field");
    check(sizeof(banHang::KhachHang) < sizeof(KhachHangChung),
          "va mo hinh nho hon han — 4 field so voi 12");

    // ---- 2. CON BUG: cung mot chu, hai nghia ----
    // Don cua chi Hoa: da chot ban (ban hang goi la "hoan tat"), nhung cong no 30 ngay
    // nen chua thu tien (ke toan KHONG goi la hoan tat).
    banHang::KhachHang daChot("KH-01", "Chi Hoa", "hoi cho", banHang::GiaiDoan::DA_CHOT);
    keToan::BenNhanHoaDon chuaThu("KH-01", "Cong ty Hoa Mai", "0301234567", "12 Le Loi", false);

    check(daChot.daHoanTat(), "BAN HANG: hoan tat = da chot -> DUNG");
    check(!chuaThu.daHoanTat(), "KE TOAN: hoan tat = da thu tien -> CHUA");
    check(daChot.daHoanTat() != chuaThu.daHoanTat(),
          "cung mot khach, cung mot chu, hai cau tra loi — va ca hai deu dung");
    // Voi mo hinh chung, `hoanTat` la MOT bool. Ai gan no? Doi nao gan thi doi kia doc
    // sai. Khong co cach va nao ngoai viec tach ra thanh hai khai niem — va tach ra thi
    // da la hai bounded context roi.
    //
    // Day la con bug dung nhu bai 81 phan 1, nhung o quy mo to chuc: o do la hai lap
    // trinh vien hieu khac nhau, o day la hai PHONG BAN hieu khac nhau. Va ho deu dung.

    // ---- 3. NOI HAI NGU CANH BANG MA, KHONG BANG OBJECT ----
    check(daChot.maKhach == chuaThu.maKhach, "cung mot con nguoi ngoai doi");
    // Nhung KHONG co duong nao di tu `banHang::KhachHang` sang `keToan::BenNhanHoaDon`.
    // Day chinh la bai 83 (tham chieu bang id) nang len cap do to chuc: hai ngu canh chia
    // se mot DINH DANH, khong chia se mot MO HINH.
    //
    // Va trinh bien dich canh giup — dong duoi KHONG bien dich duoc:
    //     keToan::BenNhanHoaDon x = daChot;
    //     error: no matching function for call to 'BenNhanHoaDon::BenNhanHoaDon(banHang::KhachHang&)'
    // Chu y: dieu nay chi dung vi constructor cua `BenNhanHoaDon` khong nhan mot tham so
    // duy nhat co the chuyen doi ngam. Neu no co mot constructor mot tham so khong
    // `explicit`, C++ se tu tim duong chuyen doi — va ranh gioi vua thung mot lo.

    // ---- 4. DO CHI PHI THAY DOI ----
    // Ke toan can them `dieuKhoanThanhToan`.
    //   Mo hinh chung: sua header -> MOI file `#include` no duoc dich lai. Ba doi cung
    //                  build lai, cung test lai, cung trien khai.
    //   Tach ngu canh: sua header cua rieng ke toan -> chi cac file cua ke toan dich lai.
    int doiBiAnhHuongMoHinhChung = 3;
    int doiBiAnhHuongKhiTach = 1;
    check(doiBiAnhHuongMoHinhChung == 3 && doiBiAnhHuongKhiTach == 1, "3 doi so voi 1");
    // O C++ con so nay khong chi la chuyen lich hop: no la THOI GIAN BUILD. Mot header
    // dung chung duoc include o 400 file la 400 lan dich lai moi khi no doi mot dong —
    // va do la thu ca doi cam thay moi ngay, khac han voi "no ky thuat" truu tuong.

    // ---- 5. BAN DO NGU CANH: quan he giua cac ngu canh co TEN ----
    //
    //   Quan he            | Nghia                                   | Khi nao dung
    //   -------------------|-----------------------------------------|-------------------
    //   Doi tac            | hai doi cung doi, cung chiu trach nhiem | hai doi cung cong ty
    //   Khach/Nha cung cap | thuong nguon nghe ha nguon              | co quyen thuong luong
    //   Tuan thu           | ha nguon dung y nguyen mo hinh tren     | ben tren khong doi duoc
    //   Chong hu hong      | ha nguon DICH mo hinh tren sang cua minh| mo hinh tren xau (bai 94)
    //   Nhan chung         | hai doi cung so huu mot phan ma dung chung | rat it, rat nguy hiem
    //
    // "Nhan chung" la thu moi nguoi bat dau va hoi han: mot thu vien `common-model` ma
    // ba doi cung sua. No co moi nhuoc diem cua mo hinh chung, cong them viec khong ai
    // so huu no.
    std::map<std::string, std::string> banDo{
        {"BanHang -> KeToan", "Khach/Nha cung cap: ban hang chot don, ke toan xuat hoa don"},
        {"KeToan -> CongThue", "Tuan thu: co quan thue khong doi dinh dang vi ta"},
        {"HoTro -> BanHang", "Chong hu hong: ho tro tu dich, khong phu thuoc giai doan ban"}};
    check(banDo.size() == 3, "ban do ngu canh la tai lieu THAT, ve duoc tren mot trang giay");
    // Neu khong ve duoc ban do nay cho he thong cua ban, thi ranh gioi ngu canh chua ton
    // tai — chi co cac thu muc code cung dung chung mot mo hinh.

    // ---- 6. DICH O BIEN, MOI CHIEU MOT LAN ----
    std::vector<keToan::BenNhanHoaDon> soHoaDon;
    if (daChot.daHoanTat()) {
        soHoaDon.emplace_back(daChot.maKhach, "Cong ty Hoa Mai", "0301234567", "12 Le Loi", false);
    }
    check(soHoaDon.size() == 1, "dich o bien: mot chieu, mot cho, co ten");
    check(soHoaDon[0].maKhach == daChot.maKhach, "chi MA di qua bien");
    // Cho dich nay la noi DUY NHAT hai ngon ngu gap nhau, nen no la noi duy nhat phai sua
    // khi mot ben doi. Bai 94 noi ky ve viec bao ve minh khi ben kia co mo hinh xau.

    // ---- 7. KHI NAO KHONG TACH NGU CANH ----
    // Bounded context co chi phi that: mo hinh lap lai, ma dich o bien, du lieu dong bo
    // tre. Ba dau hieu cho thay CHUA nen tach:
    //   - Ca he thong do MOT doi lam, va moi nguoi dung cung mot bo tu ngu.
    //   - Chua tim ra duoc mot tu nao mang hai nghia (phep thu o phan 2).
    //   - So field ma moi ben phai bo trong con nho.
    // Nguoc lai, ba dau hieu DA den luc tach:
    //   - Co field ma nua so noi dung luon de rong;
    //   - Co tu ma ban phai hoi lai "y anh la hoan tat theo nghia nao";
    //   - Mot thay doi nho phai xep lich voi doi khong lien quan.
    check(true, "ranh gioi ngu canh la quyet dinh TO CHUC, duoc bieu dien bang code");

    std::cout << "OK\n";
    return 0;
}
